File: experiments/NativeDetailInterop/NativeDetailReceiver.java
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.security.MessageDigest;
import java.util.*;
import java.util.zip.DataFormatException;
import java.util.zip.Inflater;

// Plain Java receiver parity probe. No Android framework, UI, permissions or
// production endpoints. Running on a desktop JVM is not an Android device test.
public final class NativeDetailReceiver {
    private static void check(boolean value) { if (!value) throw new IllegalArgumentException("Invalid NDL1 data"); }
    private static final class Context {
        final long epoch, request; final int width, height;
        Context(long e, long r, int w, int h) {
            check(e > 0 && r > 0 && w > 0 && w <= 8192 && h > 0 && h <= 8192 && (long)w * h <= 16777216);
            epoch = e; request = r; width = w; height = h; check(count() <= 2048);
        }
        int count() { return ((width + 127) / 128) * ((height + 127) / 128); }
        int[] rect(int tile) {
            check(tile >= 0 && tile < count()); int x = tile % ((width + 127) / 128) * 128, y = tile / ((width + 127) / 128) * 128;
            return new int[] { x, y, Math.min(128, width - x), Math.min(128, height - y) };
        }
        boolean same(Context other) { return other != null && epoch == other.epoch && request == other.request && width == other.width && height == other.height; }
    }
    private static final class Header {
        final Context context; final long sequence;
        Header(byte[] data, int kind, int minimum) {
            check(data.length >= minimum && data[0] == 'N' && data[1] == 'D' && data[2] == 'L' && data[3] == '1' &&
                data[4] == kind && data[5] == 0 && data[6] == 0 && data[7] == 0);
            ByteBuffer b = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN);
            context = new Context(b.getLong(8), b.getLong(16), b.getInt(24), b.getInt(28)); sequence = b.getLong(32); check(sequence > 0);
        }
    }
    private static final class Manifest {
        final Header header; final long[] versions;
        Manifest(byte[] data) {
            header = new Header(data, 1, 44); ByteBuffer b = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN);
            int runs = b.getInt(40); check(runs > 0 && runs <= header.context.count() && data.length == 44 + runs * 10);
            versions = new long[header.context.count()]; int tile = 0;
            for (int i = 0; i < runs; i++) {
                int count = b.getShort(44 + i * 10) & 65535; long version = b.getLong(46 + i * 10);
                check(count > 0 && count <= versions.length - tile && version > 0 && version <= header.sequence);
                Arrays.fill(versions, tile, tile + count, version); tile += count;
            }
            check(tile == versions.length);
        }
    }
    private static final class Chunk {
        final Header header; final int tile, total, offset, chunkBytes; final long version; final byte[] digest, data;
        Chunk(byte[] input) {
            check(input.length >= 96 && (input[4] == 2 || input[4] == 3));
            chunkBytes = input[4] == 2 ? 4096 : 512;
            header = new Header(input, input[4], 96); ByteBuffer b = ByteBuffer.wrap(input).order(ByteOrder.LITTLE_ENDIAN);
            tile = b.getInt(40); version = b.getLong(44); total = b.getInt(52); offset = b.getInt(56); int size = b.getInt(60);
            header.context.rect(tile);
            check(version > 0 && version <= header.sequence && total > 0 && total <= 96 * 1024 && offset >= 0 && offset < total && offset % chunkBytes == 0);
            check(size == Math.min(chunkBytes, total - offset) && input.length == 96 + size);
            digest = Arrays.copyOfRange(input, 64, 96); data = Arrays.copyOfRange(input, 96, input.length);
        }
    }
    private static final class Assembly {
        final Chunk header; final long started; final byte[] data; final boolean[] received;
        Assembly(Chunk h, long now) { header = h; started = now; data = new byte[h.total]; received = new boolean[(h.total + h.chunkBytes - 1) / h.chunkBytes]; }
        boolean matches(Chunk h) { return header.header.sequence == h.header.sequence && header.version == h.version &&
            header.total == h.total && header.chunkBytes == h.chunkBytes && MessageDigest.isEqual(header.digest, h.digest); }
    }
    private static final class Patch {
        final long version; final byte[] rgba;
        Patch(long v, byte[] pixels) { version = v; rgba = pixels; }
    }
    private static final class Cache {
        Context context; Manifest displayed; boolean enabled; int[] viewport;
        long minimum, now;
        final Map<Integer, Assembly> assemblies = new HashMap<>();
        final LinkedHashMap<Integer, Patch> patches = new LinkedHashMap<>();
        void clear() { assemblies.clear(); patches.clear(); }
        String reset(long[] p) {
            check(p.length == 8);
            for (int i = 2; i < p.length; i++) check(p[i] >= 0 && p[i] <= Integer.MAX_VALUE);
            Context next = new Context(p[0], p[1], (int)p[2], (int)p[3]);
            check(p[4] >= 0 && p[5] >= 0 && p[6] > 0 && p[7] > 0 && p[4] + p[6] <= next.width && p[5] + p[7] <= next.height);
            check(context == null || next.epoch > context.epoch || (next.epoch == context.epoch && next.request > context.request));
            clear(); context = next; viewport = new int[] {(int)p[4], (int)p[5], (int)p[6], (int)p[7]}; enabled = true; displayed = null; minimum = 0;
            return "reset";
        }
        String present(byte[] data) {
            Manifest m = new Manifest(data);
            if (!enabled || !m.header.context.same(context) || (displayed != null && m.header.sequence <= displayed.header.sequence)) return "rejected";
            if (displayed != null) for (int tile = 0; tile < context.count(); tile++) if (m.versions[tile] < displayed.versions[tile]) return "rejected";
            patches.entrySet().removeIf(entry -> entry.getValue().version != m.versions[entry.getKey()]);
            assemblies.entrySet().removeIf(entry -> entry.getValue().header.version != m.versions[entry.getKey()]);
            displayed = m; return "presented";
        }
        String off() { clear(); enabled = false; displayed = null; return "ok"; }
        String interaction() {
            if (displayed != null && displayed.header.sequence == Long.MAX_VALUE) return off();
            clear(); minimum = displayed == null ? 1 : displayed.header.sequence + 1; return "ok";
        }
        String receive(byte[] bytes, long time) throws Exception {
            check(time >= now); now = time;
            assemblies.entrySet().removeIf(entry -> time - entry.getValue().started > 1500);
            if (!enabled || displayed == null) return "Inactive";
            Chunk c;
            try { c = new Chunk(bytes); } catch (IllegalArgumentException ex) { return "Invalid"; }
            if (!c.header.context.same(context) || c.header.sequence < minimum) return "Stale";
            if (c.header.sequence > displayed.header.sequence) return "Future";
            if (c.version != displayed.versions[c.tile]) return "Stale";
            int[] r = context.rect(c.tile);
            if (!(r[0] < viewport[0] + viewport[2] && viewport[0] < r[0] + r[2] && r[1] < viewport[1] + viewport[3] && viewport[1] < r[1] + r[3])) return "OutsideViewport";
            if (patches.containsKey(c.tile)) return "Duplicate";
            Assembly a = assemblies.get(c.tile);
            if (a == null) {
                if (assemblies.size() >= 2) return "Limited";
                a = new Assembly(c, time); assemblies.put(c.tile, a);
            }
            if (!a.matches(c)) {
                if (c.header.sequence < a.header.header.sequence) return "Stale";
                if (c.header.sequence > a.header.header.sequence && c.offset == 0) { a = new Assembly(c, time); assemblies.put(c.tile, a); }
                else return "Invalid";
            }
            int index = c.offset / c.chunkBytes;
            if (a.received[index]) {
                if (Arrays.equals(Arrays.copyOfRange(a.data, c.offset, c.offset + c.data.length), c.data)) return "Duplicate";
                assemblies.remove(c.tile); return "Invalid";
            }
            System.arraycopy(c.data, 0, a.data, c.offset, c.data.length); a.received[index] = true;
            for (boolean received : a.received) if (!received) return "Partial";
            assemblies.remove(c.tile);
            if (!MessageDigest.isEqual(hash(a.data), c.digest)) return "Invalid";
            byte[] rgba;
            try { rgba = inflate(a.data, r[2] * r[3] * 4); } catch (IllegalArgumentException | DataFormatException ex) { return "Invalid"; }
            while (patches.size() >= 64) patches.remove(patches.keySet().iterator().next());
            patches.put(c.tile, new Patch(c.version, rgba)); return "Applied";
        }
        int pendingBytes() { int bytes = 0; for (Assembly a : assemblies.values()) bytes += a.data.length; return bytes; }
        String fingerprint() throws Exception {
            StringBuilder text = new StringBuilder();
            for (Map.Entry<Integer, Patch> entry : new TreeMap<>(patches).entrySet()) {
                if (text.length() != 0) text.append('\n');
                text.append(entry.getKey()).append('|').append(entry.getValue().version).append('|').append(hex(hash(entry.getValue().rgba)));
            }
            return hex(hash(text.toString().getBytes(StandardCharsets.US_ASCII)));
        }
    }
    private static byte[] inflate(byte[] compressed, int expected) throws DataFormatException {
        check(expected > 0 && expected <= 65536 && compressed.length <= 96 * 1024);
        Inflater inflater = new Inflater();
        try {
            inflater.setInput(compressed); byte[] result = new byte[expected + 1]; int count = 0;
            while (!inflater.finished() && count < result.length) {
                int read = inflater.inflate(result, count, result.length - count);
                if (read == 0) break;
                count += read;
            }
            check(count == expected && inflater.finished() && inflater.getRemaining() == 0); return Arrays.copyOf(result, expected);
        } finally { inflater.end(); }
    }
    private static byte[] hash(byte[] bytes) throws Exception { return MessageDigest.getInstance("SHA-256").digest(bytes); }
    private static String hex(byte[] bytes) {
        StringBuilder text = new StringBuilder(); for (byte value : bytes) text.append(String.format(Locale.ROOT, "%02X", value & 255)); return text.toString();
    }
    public static void main(String[] args) throws Exception {
        if (args.length != 1) throw new IllegalArgumentException("Usage: NativeDetailReceiver <interop-vectors.txt>");
        Cache cache = new Cache(); int count = 0;
        for (String line : Files.readAllLines(Paths.get(args[0]), StandardCharsets.UTF_8)) {
            if (line.startsWith("\uFEFF")) line = line.substring(1);
            String[] row = line.split("\t", -1); check(row.length == 6); long now = Long.parseLong(row[1]); String result;
            try {
                switch (row[0]) {
                    case "reset": result = cache.reset(Arrays.stream(row[2].split(",")).mapToLong(Long::parseLong).toArray()); break;
                    case "frame": result = cache.present(Base64.getDecoder().decode(row[2])); break;
                    case "chunk": result = cache.receive(Base64.getDecoder().decode(row[2]), now); break;
                    case "input": result = cache.interaction(); break;
                    case "off": result = cache.off(); break;
                    default: throw new IOException("Unknown test operation");
                }
            } catch (IllegalArgumentException ex) { result = "invalid"; }
            count++;
            if (!result.equals(row[3]) || !cache.fingerprint().equals(row[4]) || cache.pendingBytes() != Integer.parseInt(row[5]))
                throw new IOException("Event " + count + " " + row[0] + ": " + result + " != " + row[3] + ", pixel/memory state mismatch");
        }
        check(count > 50);
        System.out.println("{\"passed\":true,\"events\":" + count + ",\"runtime\":\"Java\",\"scope\":\"NDL1 wire/state/RGBA parity on desktop JVM, not Android UI or device qualification\"}");
    }
}

File: experiments/NativeDetailInterop/native_detail_receiver.py
#!/usr/bin/env python3
"""NDL1 experimental receiver parity probe. Stdlib only; not a production endpoint."""
import base64
import hashlib
import hmac
import json
from pathlib import Path
import struct
import sys
import zlib


def checked(condition):
    if not condition:
        raise ValueError("Invalid NDL1 data")


def context(value):
    epoch, request, width, height = value
    checked(0 < epoch < 2**63 and 0 < request < 2**63 and 0 < width <= 8192 and 0 < height <= 8192)
    checked(width * height <= 16_777_216 and tiles(value) <= 2048)
    return tuple(value)


def tiles(ctx):
    return ((ctx[2] + 127) // 128) * ((ctx[3] + 127) // 128)


def tile_rect(ctx, tile):
    checked(0 <= tile < tiles(ctx))
    x, y = tile % ((ctx[2] + 127) // 128) * 128, tile // ((ctx[2] + 127) // 128) * 128
    return x, y, min(128, ctx[2] - x), min(128, ctx[3] - y)


def header(data, kind, minimum):
    checked(len(data) >= minimum and data[:4] == b"NDL1" and data[4:8] == bytes((kind, 0, 0, 0)))
    ctx = context(struct.unpack_from("<qqii", data, 8))
    seq, = struct.unpack_from("<q", data, 32)
    checked(seq > 0)
    return ctx, seq


def manifest(data):
    ctx, seq = header(data, 1, 44)
    runs, = struct.unpack_from("<i", data, 40)
    checked(0 < runs <= tiles(ctx) and len(data) == 44 + 10 * runs)
    versions = []
    for index in range(runs):
        count, version = struct.unpack_from("<Hq", data, 44 + 10 * index)
        checked(0 < count <= tiles(ctx) - len(versions) and 0 < version <= seq)
        versions.extend([version] * count)
    checked(len(versions) == tiles(ctx))
    return ctx, seq, tuple(versions)


def chunk(data):
    checked(len(data) >= 96 and data[4] in (2, 3))
    chunk_bytes = 4096 if data[4] == 2 else 512
    ctx, seq = header(data, data[4], 96)
    tile, version, total, offset, size = struct.unpack_from("<iqiii", data, 40)
    tile_rect(ctx, tile)
    checked(0 < version <= seq and 0 < total <= 96 * 1024 and 0 <= offset < total and offset % chunk_bytes == 0)
    checked(size == min(chunk_bytes, total - offset) and len(data) == 96 + size)
    return ctx, seq, tile, version, total, offset, data[64:96], data[96:], chunk_bytes


class Cache:
    def __init__(self):
        self.ctx = None
        self.enabled = False
        self.displayed = None
        self.viewport = None
        self.minimum = 0
        self.now = 0
        self.clear()

    def clear(self):
        self.assemblies = {}
        self.patches = {}  # Insertion order is the bounded FIFO eviction order.

    def reset(self, values):
        ctx = context(values[:4])
        x, y, w, h = viewport = tuple(values[4:])
        checked(x >= 0 and y >= 0 and w > 0 and h > 0 and x + w <= ctx[2] and y + h <= ctx[3])
        checked(self.ctx is None or ctx[0] > self.ctx[0] or (ctx[0] == self.ctx[0] and ctx[1] > self.ctx[1]))
        self.clear()
        self.ctx, self.viewport, self.enabled, self.displayed, self.minimum = ctx, viewport, True, None, 0
        return "reset"

    def present(self, data):
        item = manifest(data)
        if not self.enabled or item[0] != self.ctx or (self.displayed and item[1] <= self.displayed[1]):
            return "rejected"
        if self.displayed and any(new < old for new, old in zip(item[2], self.displayed[2])):
            return "rejected"
        self.patches = {tile: value for tile, value in self.patches.items() if value[0] == item[2][tile]}
        self.assemblies = {tile: value for tile, value in self.assemblies.items() if value[0][3] == item[2][tile]}
        self.displayed = item
        return "presented"

    def off(self):
        self.clear()
        self.enabled, self.displayed = False, None
        return "ok"

    def interaction(self):
        if self.displayed and self.displayed[1] == 2**63 - 1:
            return self.off()
        self.clear()
        self.minimum = self.displayed[1] + 1 if self.displayed else 1
        return "ok"

    def receive(self, data, now):
        checked(now >= self.now)
        self.now = now
        self.assemblies = {tile: value for tile, value in self.assemblies.items() if now - value[1] <= 1500}
        if not self.enabled or not self.displayed:
            return "Inactive"
        try:
            item = chunk(data)
        except (ValueError, struct.error):
            return "Invalid"
        ctx, seq, tile, version, total, offset, digest, content, chunk_bytes = item
        if ctx != self.ctx or seq < self.minimum:
            return "Stale"
        if seq > self.displayed[1]:
            return "Future"
        if version != self.displayed[2][tile]:
            return "Stale"
        x, y, w, h = tile_rect(ctx, tile)
        vx, vy, vw, vh = self.viewport
        if not (x < vx + vw and vx < x + w and y < vy + vh and vy < y + h):
            return "OutsideViewport"
        if tile in self.patches:
            return "Duplicate"
        if tile not in self.assemblies:
            if len(self.assemblies) >= 2:
                return "Limited"
            self.assemblies[tile] = [item, now, bytearray(total), set()]
        entry = self.assemblies[tile]
        previous = entry[0]
        if (previous[1], previous[3], previous[4], previous[6], previous[8]) != (seq, version, total, digest, chunk_bytes):
            if seq < previous[1]:
                return "Stale"
            if seq > previous[1] and offset == 0:
                entry = [item, now, bytearray(total), set()]
                self.assemblies[tile] = entry
            else:
                return "Invalid"
        if offset in entry[3]:
            if entry[2][offset:offset + len(content)] == content:
                return "Duplicate"
            del self.assemblies[tile]
            return "Invalid"
        entry[2][offset:offset + len(content)] = content
        entry[3].add(offset)
        if len(entry[3]) != (total + chunk_bytes - 1) // chunk_bytes:
            return "Partial"
        del self.assemblies[tile]
        if not hmac.compare_digest(hashlib.sha256(entry[2]).digest(), digest):
            return "Invalid"
        try:
            decoder = zlib.decompressobj()
            rgba = decoder.decompress(entry[2], w * h * 4 + 1)
            checked(len(rgba) == w * h * 4 and decoder.eof and not decoder.unconsumed_tail and not decoder.unused_data)
        except (ValueError, zlib.error):
            return "Invalid"
        while len(self.patches) >= 64:
            del self.patches[next(iter(self.patches))]
        self.patches[tile] = version, rgba
        return "Applied"

    def fingerprint(self):
        text = "\n".join(f"{tile}|{version}|{hashlib.sha256(rgba).hexdigest().upper()}"
                         for tile, (version, rgba) in sorted(self.patches.items()))
        return hashlib.sha256(text.encode("ascii")).hexdigest().upper()

    def pending_bytes(self):
        return sum(len(entry[2]) for entry in self.assemblies.values())


def run(path):
    cache = Cache()
    count = 0
    for line in Path(path).read_text(encoding="utf-8-sig").splitlines():
        operation, now, data, expected, fingerprint, pending = line.split("\t")
        try:
            if operation == "reset":
                result = cache.reset(tuple(map(int, data.split(","))))
            elif operation == "frame":
                result = cache.present(base64.b64decode(data, validate=True))
            elif operation == "chunk":
                result = cache.receive(base64.b64decode(data, validate=True), int(now))
            elif operation == "input":
                result = cache.interaction()
            elif operation == "off":
                result = cache.off()
            else:
                raise RuntimeError("Unknown test operation")
        except (ValueError, struct.error):
            result = "invalid"
        count += 1
        if result != expected or cache.fingerprint() != fingerprint or cache.pending_bytes() != int(pending):
            raise AssertionError(f"event {count} {operation}: {result} != {expected}, cache/pending parity={cache.fingerprint() == fingerprint}/{cache.pending_bytes() == int(pending)}")
    checked(count > 50)
    print(json.dumps(dict(passed=True, events=count, runtime="Python", scope="NDL1 wire/state/RGBA parity; not a production Linux viewer or GUI test")))


if __name__ == "__main__":
    if len(sys.argv) != 2:
        raise SystemExit("Usage: native_detail_receiver.py <interop-vectors.txt>")
    run(sys.argv[1])
